mr/src/main/java/org/apache/iceberg/mr/IcebergInputFormat.java

```diff
@@ -0,0 +1,345 @@
+import org.apache.iceberg.TableScan;
+import org.apache.iceberg.avro.Avro;
+import org.apache.iceberg.catalog.Catalog;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.common.DynClasses;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.hadoop.HadoopInputFile;
+import org.apache.iceberg.hadoop.HadoopTables;
+import org.apache.iceberg.hive.HiveCatalogs;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.orc.ORC;
+import org.apache.iceberg.parquet.Parquet;
+import org.apache.iceberg.types.TypeUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class IcebergInputFormat<T> extends InputFormat<Void, T> {
+  private static final Logger LOG = LoggerFactory.getLogger(IcebergInputFormat.class);
+
+  static final String FILTER_EXPRESSION = "iceberg.mr.filter.expression";
+  static final String TABLE_SCHEMA = "iceberg.mr.table.schema";
+  static final String TABLE_PATH = "iceberg.mr.table.path";
+  static final String READ_SCHEMA = "iceberg.mr.read.schema";
+  static final String READ_SUPPORT = "iceberg.mr.read.support";
+
+  private transient Table table;
+  private transient List<InputSplit> splits;
+
+  public IcebergInputFormat() {
+  }
+
+  @Override
+  public List<InputSplit> getSplits(JobContext context) {
+    if (splits != null) {
+      LOG.info("Returning cached splits: {}", splits.size());
+      return splits;
+    }
+
+    Configuration conf = context.getConfiguration();
+    table = getTable(conf);
+    TableScan scan = table.newScan();
+    //TODO add caseSensitive, snapshot id etc..
+
+    Expression filterExpression = SerializationUtil.deserializeFromBase64(conf.get(FILTER_EXPRESSION));
+    if (filterExpression != null) {
+      scan = scan.filter(filterExpression);
+    }
+
+    final String schemaStr = conf.get(READ_SCHEMA);
+    if (schemaStr != null) {
+      // Not sure if this is having any effect?
+      scan.project(SchemaParser.fromJson(schemaStr));
+    }
+
+    splits = Lists.newArrayList();
+    try (CloseableIterable<CombinedScanTask> tasksIterable = scan.planTasks()) {
+      tasksIterable.forEach(task -> splits.add(new IcebergSplit(task)));
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to close table scan: %s", scan);
+    }
+
+    return splits;
+  }
+
+  @Override
+  public RecordReader<Void, T> createRecordReader(InputSplit split, TaskAttemptContext context) {
+    return new IcebergRecordReader();
+  }
+
+  public static ConfBuilder updateConf(
+      Configuration conf, String path, Class<? extends ReadSupport<?>> readSupportClass) {
+    return new ConfBuilder(conf, path, readSupportClass);
+  }
+
+  public static class ConfBuilder {
+    private final Configuration conf;
+
+    public ConfBuilder(Configuration conf, String path, Class<? extends ReadSupport<?>> readSupportClass) {
```

Comment:
If we are going with the builder pattern, then let's set the table/path and in-memory data model in builder methods instead of here. I prefer not to minimize constructor args with builders by removing anything that might not need to be set (e.g. use Iceberg generics by default).

Reply:
in-memory data model +1.
For table path, which is mandatory, does it make more sense in the constructor? Then users cannot omit it.  Granted that if they omit it, the code will just fail, but I general add required args as const. params and options and methods to the builder.

Reply:
Decided to set everything as builder methods except configuration, similar to Kite